joomla/joomla-cms
# [4.0] Subform UX

Pull Request for Issue #28389.

### Summary of Changes
Joomla! has a custom field type called **Subfields** and recently, renamed to **Subform** in a different PR. From it's documentation [https://docs.joomla.org/Custom_fields_type:_Subfields](https://docs.joomla.org/Custom_fields_type:_Subfields)

> the aim of it is to provide the user the possibility to create a repeatable custom field for an item with multiple related fields

There are some UX issues with that kind of custom field as described below (see https://github.com/joomla/joomla-cms/issues/28389 for detailed discussion):

- On Fields Management screen, there is no way to know if a custom field is a standard field or it is a subfield (part of a **Subform** custom field type).
- To prevent a custom field from being used both as standard field and as part of a **Subform**, we need to set **Category** of that custom field to **None**. It's hard to know about that.
- There is no filter option to allow seeing custom fields which are being used as part of **Subform**.
- There is no option to limit maximum number of rows which can be added when adding data for a **Subform**. 
- Users won't know that they must add some standard fields before they can add a **Subform**.

This PR tries to solve these issues base on suggestions from @AndySDH. See https://github.com/joomla/joomla-cms/issues/28389#issuecomment-816878826 for details (one of the suggestion is renaming the custom field type from Subfields to Subform but it is handled by different PR) 

1. Add a new setting **Use In Subform** on add/edit custom field screen to indicate that the field will be used in **Subform**.

- When it set to Yes, categories selection will become hidden.
- Category of the field will become **None**.

2. Add a filter to search tool to allow filtering fields belong to Subform (Yes/No). By default, all fields will be show.

<img width="893" alt="filter" src="https://user-images.githubusercontent.com/977664/114297720-7f0d6900-9adc-11eb-98e1-dbed55df5703.png">

3. If you add a **Subform** custom field but forgot to create at least one standard field before , there will be a warning displayed at the top to tell you about that.
![warning](https://user-images.githubusercontent.com/977664/114297790-efb48580-9adc-11eb-82ca-5dc5d85b1165.png)

4. On **Custom Fields Management** screen, add a small badge next to the field so that we know if it is a field which is used for **SubForm Only**.

![subform_only](https://user-images.githubusercontent.com/977664/114305273-a296da80-9b01-11eb-9331-b269f447a2b1.png)

5. Add a new setting called **Maximum Rows** to control maximum number of rows which you can add while adding data for that **Subform** custom field type.

### Testing Instructions
1. Install Joomla 4 nightly build. Or if you are having an existing Joomla 4 installation, please update that installation to latest Joomla 4 nightly build.
2. Download update package for this PR at https://ci.joomla.org/artifacts/joomla/joomla-cms/4.0-dev/33096/downloads/42296/Joomla_4.0.0-beta8-dev+pr.33096-Development-Update_Package.zip , go to **System -> Update -> Joomla**, upload and install that update package
3. Go to Content -> Custom Fields, create one or two **Text** custom fields. On adding custom field screen, confirm that you see a new setting called **Only Use In Subform**. Set it to **Yes** and confirm that the **Category** selection at the right is hidden. Save the field, you will be redirected to Custom Fields list screen. Confirm that you see a badge **Subform Only** displays next to that custom field (to indicate that the custom field is only used in Subform)
4. Create another custom field type **Subform** . Confirm that:
- You do not see **Only Use In Subform** setting.
- You see a new setting called **Maximum Rows**. Set it to the number you want (for example 2). In the **Fields** setting of that custom field, select some fields for it.
5. Go back to custom fields list screen, click on **Filter Options** to show search tools. You will see a new filter option **Only Use In Subform**, use it and make sure it is working as expected
6. Now, create a new article. Navigate to Fields tab, enter data for that Subform custom field which you created. Please make sure that you cannot add more rows for that custom field than the number you entered into **Maximum Number Rows** setting of that field (2 in the case)
